lib/services/unified_resource.go

```diff
@@ -998,6 +968,138 @@ type resource interface {
+	latest resource
+}
+
+func (c *singularResourceCollection) get() resource { return c.latest }
+
+func (c *singularResourceCollection) put(r resource) { c.latest = r }
+
+func (c *singularResourceCollection) remove(types.Resource) bool { return true }
+
+type resourceServerCollection struct {
+	latest  resource
+	servers map[string]struct{}
+}
+
+func (c *resourceServerCollection) get() resource {
+	return c.latest
+}
+
+func (c *resourceServerCollection) put(r resource) {
+	c.latest = r
+	if r, ok := r.(interface{ GetHostID() string }); ok {
+		c.servers[r.GetHostID()] = struct{}{}
+	}
+}
+
+func (c *resourceServerCollection) remove(r types.Resource) bool {
+	// This looks insane, but we only get a [types.ResourceHeader] in
+	// [types.OpDelete] events.
+	// The types that actually implement [resourceServer] all store the host ID
+	// in the description of the resource header metadata on deletion.
+	// If a new type is added that implements [resourceServer] and the
+	// unified resource watchers starts watching it, then please:
+	// - add it to the isResourceServer helper func
+	// - ensure host ID is stored in the metadata description in its event parser
+	// - add test coverage for it in TestUnifiedResourceWatcher_DeleteEvent
+	delete(c.servers, r.GetMetadata().Description)
+	return len(c.servers) == 0
+}
+
+func newDatabaseServerCollection(d types.DatabaseServer) *databaseServerCollection {
+	statuses := map[string]types.TargetHealthStatus{
+		d.GetHostID(): types.TargetHealthStatus(d.GetTargetHealth().Status),
+	}
+	return &databaseServerCollection{
+		aggregate: &aggregatedDatabase{
+			DatabaseServer: d,
+			status:         aggregateHealthStatuses(statuses),
+		},
+		servers: statuses,
+	}
+}
+
+type databaseServerCollection struct {
+	aggregate *aggregatedDatabase
+	servers   map[string]types.TargetHealthStatus
+}
+
+func (c *databaseServerCollection) get() resource {
+	return c.aggregate
+}
+
+func (c *databaseServerCollection) put(r resource) {
+	if r, ok := r.(types.DatabaseServer); ok {
+		c.servers[r.GetHostID()] = types.TargetHealthStatus(r.GetTargetHealth().Status)
+		c.aggregate = &aggregatedDatabase{
+			DatabaseServer: r,
+			status:         aggregateHealthStatuses(c.servers),
+		}
+	}
+}
+
+func (c *databaseServerCollection) remove(r types.Resource) bool {
+	delete(c.servers, r.GetMetadata().Description)
+	c.aggregate = &aggregatedDatabase{
+		DatabaseServer: c.aggregate.DatabaseServer,
+		status:         aggregateHealthStatuses(c.servers),
+	}
+	return len(c.servers) == 0
+}
+
+// aggregatedDatabase wraps a database server with aggregated health status.
+// This type exists to avoid cloning the resource unnecessarily, yet still
+// prevent data races.
+type aggregatedDatabase struct {
+	types.DatabaseServer
+	status types.TargetHealthStatus
+}
+
+func (d *aggregatedDatabase) GetTargetHealth() types.TargetHealth {
+	out := d.DatabaseServer.GetTargetHealth()
+	out.Status = string(d.status)
```

Comment:
Since using aggregated status, do we still need other information from `d.DatabaseServer.GetTargetHealth()`?